### DISCUSSION TITLE
CH341A futher improvement -- to support lower clock AVR chips

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1441] [Ideas]
#1215 has been merged which addressed major part of Issue #714.

* #714 
* #1215

However, #1215 only supports fast SPI mode, which means that lower clock AVR chips (including many AVR chips out of factory which will run on internal RC oscillator at around 1MHz to 1.6Mhz). It is mentioned in #714 that bitbang SPI can be used to support AVR chips with lower clock.

The other main issue with #1215 is that it only supports libusb-1.0 API. Therefore the default Windows MSVC build will not work since it uses avrdude-libusb which only supports libusb-0.1 API. This part will be tracked in a seperate issue #1339.

**Comment by mcuee** [comment]
From the following comments:
* https://github.com/avrdudes/avrdude/issues/714#issuecomment-1168047250

Reference bitbang SPI code using libusb.
https://github.com/maxxproff/avrdude-ch341/blob/main/ch341a_bitbang.c

**Comment by mcuee** [comment]
Another potential improvement is to have higher ISP clock even for the fast SPI mode.

* https://github.com/avrdudes/avrdude/pull/1215#issuecomment-1493069143

Hans mentioned in the above comment.
> It is also not possible to change the ISP clock speed. We know that it can't go lower than 1.2 MHz, but I think it should be possible to increase this. AVRs running at 20 MHz support ISP clocks up to 5 MHz.

**Comment by mcuee** [comment]
Yet another potential improvement is to add TPI support.

**Comment by mcuee** [comment]
@mariusgreuel 

When you got some time, please check if you can add MSVC build to this one. Thanks.

**Comment by stefanrueger** [comment]
Apart from the libusb/MSVC build issue this is pretty much a new and different driver, sth like `ch341_bb`. Needs to be contributed by people with an insight into the chip. There are simular bitbang driver, eg, for ftdi. So maybe the community can contribute? @Kaell?

**Comment by MCUdude** [comment]
@maxxproff hosts an Avrdude with support for ch341a bit-banging:
https://github.com/maxxproff/avrdude-ch341

**Comment by stefanrueger** [comment]
> @maxxproff hosts an Avrdude with support for ch341a bit-banging: https://github.com/maxxproff/avrdude-ch341

Great find. Making them a great candidate to submit a PR :)

**Comment by mcuee** [comment]
I have split the enhancement into two issues. This issue will mainly deal with the SPI bitbang enhancement. The MSVC build issue will be tracked in #1339.

* #1339

**Comment by mcuee** [comment]
> @maxxproff hosts an Avrdude with support for ch341a bit-banging: https://github.com/maxxproff/avrdude-ch341

He mentioned that the bit-bang driver to be this one, which was mentioned earlier.
https://github.com/maxxproff/avrdude-ch341/blob/main/ch341a_bitbang.c

**Comment by stefanrueger** [comment]
I still think this has to be driven by the community, who use the CH341A for programming, not us as maintainers. Hence, I would argue to close this and engage with a PR if and when it comes along.

**Comment by mcuee** [comment]
@stefanrueger 

The issues in github do not need to be driven by the maintainers -- the community can help to provide the PR.

Some potential contributors will first look at github to see if there are anything they can contribute. If we close issues like this, they may not know this issue at all.

**Comment by mcuee** [comment]
@stefanrueger 

I have added a `help wanted` label to show that we expect the community to help for this issue.

**Comment by mcuee** [comment]
Move this to discussion.

**Comment by mcuee** [comment]
Close this for now as it is already a known limitation documented.
https://github.com/avrdudes/avrdude/wiki/Known-limitations-of-avrdude
